[3주차] 랜선 최애 멍멍 덕질 월드컵

solar3070 opened the pull request:
## ✨ 구현 기능 명세
### **기본과제**
1. **월드컵의 진행방식은 다음과 같이 이루어져요.** ✔
    1. 4개의 참가자가 있을 경우 (4강)
        1. A vs B ***(1)*** `준결승`
        2. C vs D ***(2)*** `준결승`
        3. (1) 승리자  vs (2) 승리자 `결승`
    2. 즉 각 경기의 승리자들끼리 토너먼트를 진행하면 됩니다.
2. **각 경기마다 참가자의 View는 다음과 같이 구성해주세요.** ✔
    1. 1:1 비율로 2개의 이미지를 보여주세요.
    2. 참가자의 이름이 나타나야해요.
3. 마지막 경기의 승리자는 최종 우승자가 돼요. **그 음식에 왕관을 씌워주세요.** ✔
월드컵이 종료되는 조건을 판단하셔서 우승자를 보여주면 과제 끝!

### **심화과제**
1. 참가자 중에 승리자가 결정되면(왼쪽이던 오른쪽이던 클릭하게 되면) 해당 참가자가 승리했음을 알 수 있도록 하는 애니메이션을 구현해주세요. ✔(시간 될 때 꼭 구현합니다)
    1. 애니메이션은 자유에요. 흔들거려도 되고 커져도 되고...
    2. 제가 구현한 애니메이션은 패배한 참가자의 width를 0으로 줄이면서 transition을 주었어요.
https://github.com/THE-SOPT-WEB/leeHyeJun/pull/9
2. 결과화면에서 다시하기 버튼을 누르면 게임을 초기화해주세요. ✔
4. 새로고침할 때마다 월드컵 참가자들의 순서를 랜덤으로 바꿔주세요. ✔
    - 배열 랜덤으로 재정렬 하는 방법
        
        ```jsx
        const array = [1, 2, 3, 4, 5]
        array.sort(() => Math.random() - 0.5);
        ```
        
5. 별책부록에 있는 `내가 만든 웹 배포하기` 를 참고하셔서 배포까지 진행한 뒤 ***배포 링크***를 올려주세요. ✔


## 🎁 PR Point
- useEffect를 어떻게 사용하면 좋을지 잘 모르겠습니다. 사용하고 싶어서 이거저거 바꿔보고 난리를 쳐봤는데 몇 시간을 쏟아도 안되고.. 모르겠네요.
- styled-component 같은 경우는 jsx 파일마다 있는 게 더 좋을까요? 아님 지금처럼 style.js에 모아두는 게 더 좋을까요? 
- useRef와 useState의 사용이 좀 헷갈리네요. 뭐 렌더링이 되고 안되고 .. 이런 개념적인 것들은 알겠는데 직접 코드를 짜려니 어디에 뭐를 써야하고 이런게 어렵습니다.
- 어떻게하면 더 깔끔하게 짤 수 있을지.... 클린 코드 조언 대환영입니다. 

## 😭 어려웠던 점
- 음 뭔가 굉장히 잘못된 방향으로 구현한 것 같은 느낌? 리액트의 장점을 제대로 활용하지 못한 것 같습니다... 
- 라우터도 적용하고 컴포넌트도 좀 더 깔끔하게 나누고 싶은데 어렵네요.

## 😎 구현 결과물
배포: **https://darling-maamoul-87569f.netlify.app/** -> 이미지 로딩이 좀 느리네요

https://user-images.githubusercontent.com/63948884/167147829-a1150c4e-86c3-407a-8094-87df5c760db7.mp4

solar3070 committed:
init: 3주차 과제 초기 코드

solar3070 committed:
chore: 이상형 월드컵 이미지 추가

solar3070 committed:
chore: 게임 데이터 추가

solar3070 committed:
feat: 월드컵 기본 진행방식 구현

solar3070 committed:
refactor: Fighter 컴포넌트 분리

solar3070 committed:
feat: styled-components 적용

solar3070 committed:
design: vs 추가 및 스타일링 변경

solar3070 committed:
design: reset css 적용

solar3070 committed:
chore: style 폴더 생성

solar3070 committed:
feat: 라운드 보여주기 및 우승자 왕관

solar3070 committed:
chore: state 이름 win으로 변경

solar3070 committed:
refactor: 코드 정리

solar3070 committed:
chore: 프리티어 적용 해제

Nahee-Park commented on week3/src/components/Game.jsx:
```diff
@@ -0,0 +1,49 @@
+        <p>{win ? "우승자" : `${leftRound} / ${round.current}`}</p>
+      </Header>
+      <Main>
+        {!win && fighters.slice(0, 2).map((fighter) => 
+          (<GameItem gameInfo={fighter} handleClick={handleClick} />))}
+        {win && (<GameItem gameInfo={winners.current[0]} handleClick={handleClick} />)}
+        {win ? <img src={crown} alt="crown" /> : <img src={vs} alt="vs" />}
```
```suggestion
        {win ? (
          <>
            <GameItem gameInfo={winners.current[0]} handleClick={handleClick} />
            <img src={crown} alt="crown" />
          </>
        ) : (
          <>
            {fighters.slice(0, 2).map((fighter) => (
              <GameItem gameInfo={fighter} handleClick={handleClick} key={fighter.name} />
            ))}
            <img src={vs} alt="vs" />
          </>
        )}
```

1. 이 부분 삼항연산자를 쓴다면 이런 식으로 바꾸는 것이 좀 더 좋을 것 같습니다 ! 어차피 지금 케이스가 `win===true`일 때, `win===false`일 때 두 케이스인 것인데, 삼항연산자만으로도 충분히 포괄할 수 있을 것 같아요!
2. `map`을 쓸 때에는 꼭꼭 `key`값을 넣어줘야 합니다 ! 사실 이 이유를 확실하게 이해하려면 react의 동작방식을 좀 알아야하는데, 간단하게 설명하자면 리액트는 기존돔트리와 새로운 트리를 비교알고리즘을 통해 UI를 갱신합니다. `map`을 사용하는 경우, 보통 그 내부의 컴포넌트들이 같은 껍데기에 내용물만 달라지는 경우인데, key를 써주지 않으면 어떤 애가 같은 애고 어떤 애가 다른 애인지 리액트가 구분할 수가 없어요. (그래서 같은 노드를 효율적으로 사용할 수 있는 방식을 버리고 비효율적인 방식을 사용하게 된다고 해요ㅠ) 그래서 key값을 넣어서 얘가!! 얘다!! 하고 리액트한테 알려주는 게 중요해요! 

+) 사실 key를 넣어주지 않으면 자동적으로 리액트가 map의 index값을 넣어서 사용하게 되는데 이건 문제 생길 여지가 다분해요 (index값은 배열의 길이가 바뀌거나 하면 너무 쉽게 바뀌어버리는 값이라서요ㅠㅠ) 

좀 더 자세히 알고 싶다면 https://ko.reactjs.org/docs/reconciliation.html#the-diffing-algorithm 글을 참고해보면 좋을 것 같습니다! 

근데 사실 한 줄 요약하면 map쓸 때 꼭 key 넣어야 한다.. 입미다 ..ㅋ ㅋ ㅋ

Nahee-Park replied:
근데 추가적으로 삼항 연산자가 가독성 측면에서 또 권장하는 방식은 아니어서 .. 이 부분 로직을 새로운 컴포넌트로 빼내서 switch-case 문으로 조건부 렌더링 하는 방식이 가독성 측면에서 가장 이상적일 것 같습니다! 

그치만 이 부분은 클린코드의 영역이라 주관이 좀 개입되어 있어서.. 이런 방식도 있다 ~ 정도로만 넘어가도 좋을 것 같아여 ㅎ.ㅎ.



- Result.jsx 컴포넌트로 빼내기 ! 
```javascript
import React from 'react';
import GameItem from './GameItem';
import vs from '../assets/vs.png';
import crown from '../assets/crown.png';

function Result({ win, winners, handleClick, fighters }) {
  switch (win) {
    case true:
      return (
        <>
          <GameItem gameInfo={winners.current[0]} handleClick={handleClick} />
          <img src={crown} alt="crown" />
        </>
      );
    default:
      return (
        <>
          {fighters.slice(0, 2).map((fighter) => (
            <GameItem gameInfo={fighter} handleClick={handleClick} key={fighter.name} />
          ))}
          <img src={vs} alt="vs" />
        </>
      );
  }
}

export default Result;
```

- Game.jsx에서 이런 식으로 이용하기 !
```javascript
  return (
    <>
      <Header>
        <h1>여우 월드컵</h1>
        <p>{win ? '우승자' : `${leftRound} / ${round.current}`}</p>
      </Header>
      <Main>
        <Result win={win} winners={winners} handleClick={handleClick} fighters={fighters} />
      </Main>
    </>
  );
```

yunsun99 replied:
저도 딱 읽으면서 가독성 측면에서 아쉽다고 생각했는데 나희님이 너무 퍼펙트하게 적어놓으셔서,... 저는 개인적으로 삼항연산자 방법으로 사용한 것도 읽기에 괜찮았습니다. 나희님이 제시하신 코드 두가지 방법 모두 좋은 것 같아요.

solar3070 replied:
1. 와 삼항연산자로 저렇게 포괄하는 거 생각지도 못했네요! 근데 가독성이 떨어지는 건 확실히 맞는 것 같아서 새로운 컴포넌트로 빼는 방식으로 적용해보겠습니다. 자바스크립트에서 switch-case문은 써본 적이 없었는데 이런식으로도 사용이 가능하군요! 
2. 콘솔에서 계속 key 사용하라고 경고하던데 어디서 써야하는지 몰라서 안썼는데(ㅎ) map이였군요!! 앞으로는 꼭 key를 넣어주겠습니다!

Nahee-Park commented on week3/src/components/Game.jsx:
```diff
@@ -0,0 +1,49 @@
+    setLeftRound((current) => current - 1);
+    setFighters(fighters.slice(2));
+
+    if (fighters.length === 2) {
+      if (winners.current.length === 1) {
+        setFighters([]);
+        setWin(true);
+        return;
+      }
+      setFighters(winners.current);
+      winners.current = [];
+      round.current /= 2;
+      setLeftRound(round.current);
+    }
```
```suggestion
      if (fighters.length !== 2) return;
      if (winners.current.length === 1) {
        setFighters([]);
        setWin(true);
        return;
      }
      setFighters(winners.current);
      winners.current = [];
      round.current /= 2;
      setLeftRound(round.current);
```

이 부분 이런 식으로 수정하면 중첩 if문 탈출할 수 있어서 좀 더 깔끔할 것 같아여 헤헤... 근데 이것도 클린코드영역이라 갠취 반영임다 ㅋ ㅋ

solar3070 replied:
생각지도 못한 걸 많이 배워갑니다.. 👍 사실 이중 반복문 쓰면서도 이것밖에 방법이 없나 생각을 많이 했는데 제시해주신 방법이 더 깔끔한 것 같아서 좋네요!

yunsun99 replied:
우와...... 지나가다 진짜 이마탁 하고 지나갑니다.....

Nahee-Park commented on week3/src/components/Game.jsx:
```diff
@@ -0,0 +1,49 @@
+  const handleClick = (fighter) => {
+    winners.current.push(fighter);
+    setLeftRound((current) => current - 1);
+    setFighters(fighters.slice(2));
```
```suggestion
    setFighters(prevState => prevState.slice(2));
```
state값 업데이트할 때 state값에 직접 접근하기 보다 이런 식으로 최종 state인 prevState를 통해 업데이트 하는 편이 좋다고 합니다! ~왜냐하면  `setFighters(fighters.slice(2));` 이런 식으로 업데이트 해버리면 setFighter를 중복해서 호출하게 되는 경우가 생긴다구 해요~
setState를 중복 호출하는 경우 `setFighters(fighters.slice(2));` 내부의 fighters가 최신 값이라고 보장할 수 없는 상황이 생길 수도 있어서 PrevState를 통해 접근하는 편이 최신값을 안전하게 참조할 수 있는 방법이라구 해요 !

choichoijin replied:
> 

지나가는 길에 질문 한번 올리자면,, 혹시 왜 setFighters가 중복해서 호출되는건가요..??

KimKwon replied:
중복호출이라기보단 최신값 참조의 의미가 더 큽니다 ! !

Nahee-Park replied:
@choichoijin  어레,, 다시 읽어보니 제가 표현을 잘못한 것 같습미다.  중복해서 setState를 호출하는 상황이 생기면 저 안에 들어있는 fighters가 항상 최신값일 거라고 보장할 수 없다..라는 표현이 더 맞을 것 같아요!   팟짱님 말씀처럼  `가장 최신의 prevState를 받기 위해 저런 식으로 사용하는 편이 좋다 ! `가 맞는 것 같아여
https://www.delftstack.com/howto/react/react-setstate-prevstate/
이 아티클에서 문제 생기는 상황에 대해 자세히 다루고 있어서 한 번 읽어보시면 좋을 것 같슴다 헤헤..

solar3070 replied:
아하 setLeftRound의 경우에는 최신값을 current로 받아왔는데 setFighters에서는 그 부분을 놓친 것 같네요! 꼼꼼한 리뷰 감사합니다!

Nahee-Park reviewed (approved):
언냐 코드 구경하러 왔는데 너무 잘짜서 놀랐습미다.. 모랄까 기본적으로 개발을 되게 잘 하는 느낌 .. 갓 혜준 ㄷ ㄷ 

사실 이번 과제에서 useEffect가 필수적으로 쓰여야하는 곳이 없어서 이런 식으로 구현해도 충분히 괜찮다고 생각합니다 ! (..는 사실 저도 이번 과제에서 useEffect 안 썼어욥. ) 이번 과제에서는 상태값을 변화시키는 것이 클릭했을 때 외엔 없어서, handleClick함수에서 로직 처리를 전부 해줘도 괜찮았던 것 같아요.
useEffect같은 경우는
1. 해당 컴포넌트가 처음 렌더링 될 때 처리해줘야하는 로직이 있는 경우 (ex. 데이터 패칭 등..)
2. 특정 값이 업데이트 되었을 때 공통으로 처리해야하는 로직이 있는데 특정값이 여러군데에서 산발적으로  업데이트 되고 있는 경우
3. 해당 컴포넌트에서 더 이상 effect가 필요하지 않을 때(1,2 번을 통해 정의했던 effect를 청소해야할 때)
정도인 것 같습미다. 

는.. 사실 정말정말 필요한 시점이 오면 어쩔 수 없이 사용하게 될 겁니다 ! 긍까 꼭 써야한다!는 마음가짐을 가질 필요는 없는 것 같아여 ㅎ.ㅎ 

styled-component를 파일로 따로 빼거나, 한 파일에 쓰거나는 보통 협업할 때 컨벤션으로 정하는 영역인 것 같습니다 ! 무엇이든 조아여 사실 

useRef와 useState는 확실히 실제로 사용하려고 보면 어떻게 써야할 지 막막한 친구들인 것 같아요. 사실 거의 state를 많이 쓰긴 합니다. 보통 useRef는 돔에 접근하거나, 조용히 값을 안고 있을 때 쓰는 느낌이에요 ! 

주절주절 길었는데,, 언니 코드 리뷰하러 오겠다고 해놓고 이제야와서 ㅎ...ㅎ... 여러가지 주절주절 했는데 사실 대충 흘려도 좋씁니다. 너무 잘 짰어여 이미. 넘 고생하셨습니다 !!

solar3070 commented:
이게 바로 우수 정원사의 위엄...? 눈이 부셔서 앞이 보이질 않는다................................
대충 흘려도 좋다했지만 하나도 안흘릴거야 
박.나.히.절.대.사.랑.해
❤🧡💛💚💙💜
시간 나는 대로 바로 리팩토링 갈겨버린다

yunsun99 commented on week3/src/components/GameItem.jsx:
```diff
@@ -0,0 +1,20 @@
+import { Item, Image } from "../style/style";
+
+function GameItem(props) {
+  const { gameInfo, handleClick } = props;
```
아실 것 같지만
```
GameItem({gameInfo, handleClick}) {
...
}
```
이렇게 한줄이라도 줄일 수 있습니다...ㅎㅎㅎ

solar3070 replied:
하핳 좋아 숏코드 최고

yunsun99 commented on week3/src/style/style.js:
```diff
@@ -0,0 +1,90 @@
+  }
+`;
+
+export const Main = styled.main`
+  display: flex;
+  position: relative;
+  justify-content: center;
+
+  & > img {
+    position: absolute;
+    top: 50%;
+    left: 50%;
+    transform: translate(-50%, -50%);
+
+    width: 15rem;
+  }
+
+  & > img:nth-child(2) {
+    position: absolute;
+    top: 10%;
+    left: 50%;
+    transform: translate(-50%, -50%);
+
+    width: 10rem;
+  }
+`;
```
스타일링 코드는 일부러 한 파일 안에 몰아넣으신 걸까요? 사실 이런 부분은 개인 취향이긴 하지만 styled-components의 장점이 스타일링을 바로 그 코드파일 안에 할 수 있다는 점을 많이 들기 때문에 이렇게 따로 한 파일로 빼시는 경우에는 css를 많이 사용한다고도 들었습니다. 보통 중복되는 코드는 css에서 태그나 클래스 별로 정의해놓고 세부적인 스타일링이나 한두번 사용되는 스타일링은 styled-components를 사용해서 해당 파일 안에 적는다고 알고 있어요. 일부러 하신게 아니라면 해당 파일 안에 쓰는 방식도 한번 써보시면 좋을 것 같아요!

solar3070 replied:
오오 사실 그 부분이 고민이긴 했는데 너무 css 스타일로 생각했나봐요!! 처음 과제 낼 때 파일로 모으는 방식을 써봤으니 리팩토링할 때는 각 jsx 파일 별로 styled-component 코드를 넣어보겠습니다~ 감사합니다 :)

solar3070 committed:
refactor: Game 컴포넌트 삼항연산자 포괄
GameItem과 img

solar3070 committed:
refactor: GameResult 컴포넌트 생성
기존 Game 컴포넌트의 리턴문 로직 이동

solar3070 committed:
refactor: styled-components 각 jsx 파일로 이동

solar3070 committed:
fix: fighter가 두 명이 아닌 경우 리턴

solar3070 committed:
feat: 우승자 화면에 다시하기 버튼 추가

solar3070 committed:
design: 다시하기 버튼 스타일링

solar3070 committed:
refactor: 다시하기 버튼 styled-components 위치 변경

solar3070 committed:
fix: 우승시에만 다시하기 버튼 표시

solar3070 committed:
fix: 나왔던 데이터 또 나오는 오류 수정
gameInfo 원본 값이 수정되지 않도록 스프레드 연산자 사용 및
useState에 함수를 전달하여 지연 초기화 사용
지연 초기화: 초기값 설정은 최초 렌더링시에만 필요한데, 상태값이 바뀌어서 리렌더링될 때 불필요한 계산을 하는 것을 방지하기 위한 것

solar3070 committed:
refactor: GameResult 로직 switch-case문에서 if문으로 변경

solar3070 committed:
design: footer margin-bottom 넣기

n0eyes commented on week3/src/components/Game.jsx:
```diff
@@ -0,0 +1,114 @@
+  const winners = useRef([]);
+  const round = useRef(gameInfo.length / 2);
+  const [fighters, setFighters] = useState(() =>
+    [...gameInfo].sort(() => Math.random() - 0.5)
```
지연 초기화 굿굿

n0eyes commented on week3/src/components/Game.jsx:
```diff
@@ -0,0 +1,114 @@
+    setLeftRound((prev) => prev - 1);
+    setFighters((prev) => prev.slice(2));
+
+    if (fighters.length !== 2) return;
```
batch 때문에 useEffect 써도 좋을 것 같아~
가독성 측면 정도??

solar3070 replied:
오케이 나중에 해보겠음!!
